packages/lodestar/src/sync/utils/reputation.ts

```diff
@@ -0,0 +1,122 @@
+import {ILogger} from "@chainsafe/lodestar-utils";
+import {IReqResp} from "../../network";
+
+export async function handlePeerMetadataSequence(
+  reps: IReputationStore,
+  reqResp: IReqResp,
+  logger: ILogger,
+  peer: PeerId,
+  peerSeq: BigInt | null
+): Promise<void> {
+  const latestMetadata = reps.getFromPeerId(peer).latestMetadata;
+  if (peerSeq !== null && (!latestMetadata || latestMetadata.seqNumber < peerSeq)) {
+    try {
+      logger.verbose("Get and update metadata for peer " + peer.toB58String());
+      const metadata = await reqResp.metadata(peer);
+      updateMetadata(reps, peer.toB58String(), metadata);
+    } catch (e) {
+      logger.warn("Cannot get metadata for peer " + peer.toB58String());
+    }
+  }
+}
+
+/**
+ * Update new metadata to reputation store.
+ */
+export function updateMetadata(reps: IReputationStore, peerId: string, newMetadata: Metadata | null): void {
+  if (!reps) return;
+  const oldMetadata = reps.get(peerId).latestMetadata;
+  if (!oldMetadata) {
+    reps.get(peerId).latestMetadata = newMetadata;
+    return;
+  }
+  if (!newMetadata) return;
+  if (oldMetadata.seqNumber < newMetadata.seqNumber) reps.get(peerId).latestMetadata = newMetadata;
+}
+
+/**
+ * Find subnets that we don't have at least 1 connected peer.
+ */
+export function findMissingSubnets(reps: IReputationStore, peers: PeerId[] = []): number[] {
+  const attNets = peers
+    .map((peer) => reps.getFromPeerId(peer).latestMetadata)
+    .filter((metadata) => !!metadata)
+    .map((metadata) => {
+      return metadata ? metadata.attnets : [];
+    });
+  const missingSubnets: number[] = [];
+  for (let subnet = 0; subnet < ATTESTATION_SUBNET_COUNT; subnet++) {
+    if (!attNets.some((attNet) => attNet[subnet])) {
+      missingSubnets.push(subnet);
+    }
+  }
+  return missingSubnets;
+}
+
+/**
+ * If we're too close to maxPeer, we may want to disconnect some peers
+ * before connecting to new peers with missing subnets.
+ * @param peers current connected peers
+ * @param newPeer number of peers with new subnets to connect to
+ * @param maxPeer max peer provided as option
+ * @param reps our reputation store
+ * @param retainRatio: 0 to 1, the ratio to consider disconnecting peers, 90% by default.
+ */
+export function selectPeersToDisconnect(
+  peers: PeerId[],
+  newPeer: number,
+  maxPeer: number,
+  reps: IReputationStore,
+  retainRatio = 0.9
+): PeerId[] {
+  // only disconnect peers if we have >= 90% connected peers
+  if (peers.length < maxPeer * retainRatio) return [];
+  const numDisconnect = peers.length + newPeer - maxPeer;
+  if (numDisconnect <= 0) return [];
+  // should not disconnect important peers
+  const importantPeers = getImportantPeers(peers, reps);
+  const candidatePeers = peers.filter((peer) => !importantPeers.has(peer));
+  // worse peer on top
+  const sortedPeers = candidatePeers.sort((peer1, peer2) => {
+    const peer1Rep = reps.getFromPeerId(peer1);
+    const peer2Rep = reps.getFromPeerId(peer2);
```

Comment:
Don't want to lose peers with better supported protocols

Reply:
yeah we can do that using sync, for now only blocks_by_range and blocks_by_root is meaningful to us so we can ignore others.

I'm not sure if we have protocols for gossip peers but let's double check with @wemeetagain

Reply:
@mpetrunic seems like `DiversifyPeersBySubnetTask` should only be started after sync? Before node is synced, diversifying peers by subnet is not a concerned.